Code to expose ERSPAN on OpenStack setups

tanyatukade opened the pull request.

tanyatukade commented:
The checks are failing because the latest apicapi code is not merged and hence it is not able to convert those mos.

kentwu1 commented on aim/api/resource.py:
```diff
@@ -291,6 +291,225 @@ def __init__(self, **kwargs):
+    identity_attributes = t.identity(
+        ('name', t.name))
+    other_attributes = t.other(
+        ('admin_st', t.enum("start", "stop")),
```
What does st stands for here, status? Suggest to spell out the full name to make it clear.

tanyatukade replied:
okay will change that. It is "admin state"

kentwu1 commented on aim/db/migration/alembic_migrations/versions/26630ed17b50_erspan.py:
```diff
@@ -0,0 +1,196 @@
+        sa.UniqueConstraint('vsg_name', 'vs_name', 'src_path',
+                            name='uniq_aim_span_src_vport_identity'),
+        sa.Index('idx_aim_span_src_vport_identity', 'vsg_name',
+                 'vs_name', 'src_path'))
```
I would think we also need to specify ForeignKeyConstraint() here

kentwu1 commented on aim/tests/unit/agent/aid_universes/test_converter.py:
```diff
@@ -238,6 +238,170 @@ class TestAciToAimConverterNetflow(TestAciToAimConverterBase,
+                           'resource': 'spanVSrcGrp'}]
+    sample_input = [base.TestAimDBBase._get_example_aci_span_vsource_grp(),
+                    base.TestAimDBBase._get_example_aci_span_vsource_grp(
+                        dn='uni/infra/vsrcgrp-testSrcGrp1')]
```
We should use a different admin_st value here to cover different test case.

kentwu1 commented on aim/tests/unit/agent/aid_universes/test_converter.py:
```diff
@@ -238,6 +238,170 @@ class TestAciToAimConverterNetflow(TestAciToAimConverterBase,
+                           'resource': 'spanVSrc'}]
+    sample_input = [base.TestAimDBBase._get_example_aci_span_vsource(),
+                    base.TestAimDBBase._get_example_aci_span_vsource(
+                        dn='uni/infra/vsrcgrp-testSrcGrp/vsrc-testSrc1')]
```
We should use a different dir value here to cover different test case.

kentwu1 commented on aim/tests/unit/agent/aid_universes/test_converter.py:
```diff
@@ -3976,6 +4140,185 @@ class TestAimToAciConverterNetflow(TestAimToAciConverterBase,
+                                       base.TestAimDBBase):
+    sample_input = [
+        base.TestAimDBBase._get_example_aim_span_vsource(),
+        base.TestAimDBBase._get_example_aim_span_vsource(name='testSrc1',)]
```
We should use a different dir value here to cover different test case.

kentwu1 commented on aim/api/resource.py:
```diff
@@ -291,6 +291,225 @@ def __init__(self, **kwargs):
+        super(SpanVepgSummary, self).__init__({'monitored': False,
+                                               'dst_ip': '',
+                                               'flow_id': '1',
+                                               'ttl': '64',
```
Not sure why we put the default value for ttl=64 and mtu=1518 here.....

tanyatukade replied:
Thats because if we check on the UI and configure from the UI, those are the default values set by apic. So I used them.

kentwu1 replied:
ok

kentwu1 commented on aim/tests/unit/test_aim_manager.py:
```diff
@@ -748,6 +748,170 @@ class TestNetflowVMMExporterPolMixin(object):
+    test_required_attributes = {'name': 'testSrcGrp',
+                                'admin_st': 'start'}
+    test_search_attributes = {'name': 'testSrcGrp'}
+    test_update_attributes = {'display_name': 'OpenStack_Span_VSrc_Grp'}
```
We should add the admin_st to the test_update_attributes() also.

kentwu1 commented on aim/tests/unit/test_aim_manager.py:
```diff
@@ -748,6 +748,170 @@ class TestNetflowVMMExporterPolMixin(object):
+                                'name': 'testSrc',
+                                'dir': 'both'}
+    test_search_attributes = {'name': 'testSrc'}
+    test_update_attributes = {'display_name': 'OpenStack_Span_VSrc'}
```
We should add the dir to the test_update_attributes() also.

kentwu1 commented on aim/api/resource.py:
```diff
@@ -291,6 +291,225 @@ def __init__(self, **kwargs):
+        ('display_name', t.name),
+        ('dst_ip', t.string()),
+        ('flow_id', t.string()),
+        ('ttl', t.string()),
```
we should use t.integer for ttl and mtu field I think.

tanyatukade replied:
I tried using t.integer. But got a few errors wrt integer out of scope. hence switched back to string.

kentwu1 replied:
hmm it should work. Can you try to debug to see where that error is coming from?

tanyatukade replied:
will do

tanyatukade commented:
- Included the changes requested in the new commit. 
- Updated apicapi code is not yet merged and hence tests would fail. Will rerun once merged.
- Since I added few more resources will squash the commits after review.

tanyatukade commented:
created new PR.
